File: bottle.cpp

#include <unistd.h>
#include <iostream>

using namespace std;

enum Bottle { BeerBottle };


class Shelf {
private:    
    unsigned BottlesLeft;
public:
    Shelf( unsigned bottlesbought ) : BottlesLeft( bottlesbought )
    {}
    void TakeOneDown()
    {
        if (!BottlesLeft)
            throw BeerBottle;
        BottlesLeft--;
    }
    operator int () {
        return BottlesLeft;
    }
};


int main( int, char ** )
{
    Shelf Beer(99);
    try {
        for (;;) {
            char const* plural = (int)Beer !=1 ? "s" : "";
            cout << (int)Beer << " bottle" << plural << " of beer on the wall. Take one down." << endl;
            Beer.TakeOneDown();
            cout << (int)Beer << " bottle" << plural << " of beer on the wall." << endl;
        }

    } catch ( Bottle ) {
        cout << "No more bottles" << endl;
    }

    sleep(20);
    return 0;
}

File: bottle.rb
# TODO refactoing
def plural(n)
	"bottle"+(n==1&&''||'s')
end
def word(n)
	(n==1&&"it"||"one")
end
def beer(n,is_first,start_at)
	if n<0 then
		return
	elsif n==0 then
		return "no more bottles of beer on the wall.\n\nNo more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, #{start_at} bottles of beer on the wall."
	else
		pl = plural(n)
		wd = word(n)
		if is_first then
			start_at = n
			return "#{n} #{pl} of beer on the wall, #{n} #{pl} of beer.\nTake #{wd} down and pass it around, "+beer(n-1,false,start_at)
		else
			return "#{n} #{pl} of beer on the wall.\n\n#{n} #{pl} of beer on the wall, #{n} #{pl} of beer.\nTake #{wd} down and pass it around, "+beer(n-1,false,start_at)
		end
	end
end

puts beer(99,true,-1)

sleep(20)